include/lbann/utils/impl/rocm.hpp

```diff
@@ -0,0 +1,254 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2019, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#include <thrust/system/hip/execution_policy.h>
+
+// Headers for HIP
+#ifdef __HIPCC__
+#ifdef HYDROGEN_HAVE_CUB
+#include "hipcub/block/block_reduce.hpp"
+#endif // HYDROGEN_HAVE_CUB
+#include <limits>
+#include <hip/hip_fp16.h>
+#endif // __HIPCC__
+
+namespace lbann {
+
+// -------------------------------------------------------------
+// Device functions
+// -------------------------------------------------------------
+#ifdef __HIPCC__
+
+// Atomic add function
+__device__ __forceinline__
+__half gpu_lib::atomic_add(__half* address, __half val) {
+  unsigned int* address_as_uint = (unsigned int*) address;
+  unsigned int old = *address_as_uint;
+  __half* old_as_half = (__half*) &old;
+  unsigned int assumed;
+  unsigned int updated;
+  __half* updated_as_half = (__half*) &updated;
+  do {
+    assumed = old;
+    updated = old;
+    *updated_as_half += val;
+    old = atomicCAS(address_as_uint, assumed, updated);
+  } while (assumed != old);
+  return *old_as_half;
+}
+__device__ __forceinline__
+float gpu_lib::atomic_add(float* address, float val) {
+  return atomicAdd(address, val);
+}
+__device__ __forceinline__
+double gpu_lib::atomic_add(double* address, double val) {
+  return atomicAdd(address, val);
+}
```

Comment:
As far as I can tell, ROCm doesn't support `atomicAdd` for `double`. If we want full compatibility, we can copy the implementation from https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomic-functions.